### PR TITLE
Resolve #299 - Untargettable buildings

### LINKED ISSUE #299: Attacking buildings that are supposed to be invulnerable
We can attack nexus, inhibitors, and towers that are supposed to be untargetable and invulnerable, including fountain turret. Fix it.

### PR DESCRIPTION
**Pull request by Kubeer**
*Buildings now can be destroyed only in the correct order
*Untargettable buildings also have 99999 armor and magic resistance

Refs #299

### CODE REVIEW

**Comment by danil179** [comment]
Why do we need the protection if the buildings are untargettable?

File: GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs [hunk 1]
```diff
@@ -94,6 +137,39 @@ public override void Update(float diff)
                 _game.PacketNotifier.NotifySetTarget(this, null);
             }
 
+            if (_dependOnAll != null || _dependOnSingle != null)
```
Comment (danil179) [inline GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs]: All this part can go to a function `UpdateProtection`.

**Comment by Kubeer** [comment]
In issue report I saw that they have to be invulnerable so just to make sure I added this.

**Comment by danil179** [comment]
> 
> 
> In issue report I saw that they have to be invulnerable so just to make sure I added this.

But is it essential? If not, please remove this part as it clutters the code and complicate things.

**Comment by danil179** [comment]
I think that this solution is not quite good. You should add something like "ProtectionManager" instance to each building that will get and handle all of these protection mechanisms and will avoid the repetitive code and the too much arguments in the constructors. I also  was thinking about doing class "ProtectedBuilding", but this will create a lot of problems so please don't try to solve it with inheritances.

**Comment by lzardy** [comment]
> I think that this solution is not quite good. You should add something like "ProtectionManager" instance to each building that will get and handle all of these protection mechanisms and will avoid the repetitive code and the too much arguments in the constructors. I also was thinking about doing class "ProtectedBuilding", but this will create a lot of problems so please don't try to solve it with inheritances.

Having a central ProtectionManager would also be useful for invulnerability for disconnected players.

**Comment by Kubeer** [comment]
Ok, so I'll make a ProtectionManager but should i commit it here or make a new branch instead?

**Comment by lzardy** [comment]
> Ok, so I'll make a ProtectionManager but should i commit it here or make a new branch instead?

Here. You can make a new PR for replacing the current invuln. for disconnected players with ProtectionManager.

**Comment by xxx3456789** [comment]
if disconnect then apply invulnerability 
and edit something in Damage or Buff be like
if this object have invulnerability, then "TakeDamage" 's damage value will be 0 (its conception so only take for example)

**Comment by Kubeer** [comment]
Maybe some functions needs renaming but I couldn't came up with something better than this.

**Review by danil179 — APPROVED** [review]
Approved.

**Pull request merged**